### DISCUSSION TITLE
Step pulse "Jitter" on some processor

### DISCUSSION THREAD

**Discussion by fireup924** [terjeio/grblHAL#275] [Q&A]
I've tested grblHAL on the following processors and measured the timing between pulses and notice some have more "jitter" than others (timing between consecutive pulses varies, while average frequency is stable)

IMXRT1062 (very little or no jitter)
RP2040 (very little or no jitter)
STM32F103 (some jitter, but decent)
STM32F401 (noticeable jitter)
STM32F446 (noticeable jitter)
ESP32 (very bad jitter)

Was wondering why are jitter on some processor so bad such as EPS32? Can they be improved somehow?

**Comment by phil-barrett** [comment]
I don't have an answer for you as I have done most of my testing on iMXRT1062. I did one accumulation test a while ago with my O'scope in persistence mode and saw a pretty solid signal.  Jitter would show up as faint traces outside the square wave.

How are you measuring it? What is your set up? What level of jitter are you seeing?  And, what constitutes an acceptable level of jitter? 

While in a lot systems, like for audio, even a little jitter is considered bad. But for motion control, some level of jitter is going to be pretty meaningless because of the inertia of motors and the physical systems they drive. As long as the long term average (over seconds?) frequency is the intended frequency, I don't think there will be much actual error.  Perhaps lasers would show this more readily due to lower mass of the motion components?

Also, I wonder if acceleration is having some effect here.

[edit]  I wonder if Bart's stuff (based on ESP32) has jitter too?

**Reply by fireup924** [reply]
I'm measuring using O'scope and logic analyzer probe from Saleae. On the ESP32 using grblHAL the pulse between every other pulse is stable, but consecutive pulses floats around.

I've tested with Bart's stuff, but I have other issue with his stuff, his pulse train is cleaner, but the status update rate lag's for me with the Gcode sender I'm using.

**Reply by phil-barrett** [reply]
I have been seeing a little jitter with my logic analyzer (saleae logic 16, iirc) but only when I am at the limit of the resolution so hard to say if it is aliasing or true jitter.

**Reply by terjeio** [reply]
> his pulse train is cleaner

Interesting, which step generation mode are you using for Bart's port? grblHAL uses RMT.

Jitter is due to interrupt latency, most processors require two (or three if delayed) interrupts to be processed for each step pulse generated. RP2040 and ESP32 only one due to step pulse generation beeing handled off CPU.

STM32F4xx showing more jitter than F103 is a bit strange - maybe I have set interrupt priority wrong or lazy task switching is disabled?

ESP32 is running grbl as a FreeRTOS task, this may explain why this has more jitter. Code may also be fetched from external flash before execution, this may affect jitter as well. Documentation for this is hard to find - it was my intention to check this out but I never got around to it. Worst case this may cause jitter up to 300 microseconds (IIRC), which is the time it takes to fetch a page of code from flash. Interrupt code can be placed in RAM by decorating relevant functions with `IRAM_ATTR`, IMO Bart's port are missing some of these so I would expect his port to have more jitter...

**Reply by fireup924** [reply]
I tested again the latest code on a F446 and it seems the jitter has gone away for this processor. The code I tested 3-4 months ago did have jitter, now it's as good as RP2040. I also had issues with LPC1769 in the pass (BTT SKR V1.4 Turbo) with jitter and will try this latest version on it to see if it gets better as well.

**Reply by terjeio** [reply]
@fireup924 It would be nice if you specify which version of the driver, which board and if native USB or UART mode is used when testing.

The STM32F303 driver had a lot of jitter when using native USB due to the USB interrupt having having top priority and taking a long time to complete. Reducing the priority to below the stepper and step timer interrupts (in the ST driver layer) resolved that.

I have a F411 Blackpill that I want to check for the same behaviour as the F303 but I am unable to get USB working even if the code is nearly similar to the F303 code. The port shows up but I cannot connect to it unless I remove/change code in the grblHAL USB serial layer. The removed code does not even get executed when it fails, perhaps due to corrupted memory? Since the F411 and F401 shares the same code I would like to know if it is a Blackpill and if you use native USB to connect. If so then there is something seriously wrong with the F411 chips I have.BTW I have two F411 Blackpills, both behave the same and are a chinese knock-offs...
